src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -35,73 +36,62 @@ internal abstract class AbstractBraceCompletionService : IBraceCompletionService
         /// </summary>
         protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
 
-        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+        public abstract bool AllowOverType(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
 
-        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        public Task<bool> HasBraceCompletionAsync(BraceCompletionContext context, Document document, CancellationToken cancellationToken)
         {
-            var closingPoint = braceCompletionContext.ClosingPoint;
-            if (closingPoint < 1)
-                return null;
-
-            var openingPoint = braceCompletionContext.OpeningPoint;
-            var document = braceCompletionContext.Document;
-
-            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-            if (sourceText[openingPoint] != OpeningBrace)
-                return null;
```

Comment:
where did this logic go?

Reply:
Extracted a couple of helpers to `BraceCompletionContext`.